src/Microsoft.AspNetCore.Sockets.Client.Http/Internal/HttpConnectionSession.cs

```diff
@@ -0,0 +1,509 @@
+                throw;
+            }
+
+            // if the connection is not in the Connecting state here it means the user called DisposeAsync while
+            // the connection was starting
+            if (ChangeState(from: ConnectionState.Connecting, to: ConnectionState.Connected) == ConnectionState.Connecting)
+            {
+                _closeTcs = new TaskCompletionSource<object>();
+
+                _ = Input.Completion.ContinueWith(async t =>
+                {
+                    // Grab the exception and then clear it.
+                    // See comment at AbortAsync for more discussion on the thread-safety
+                    var abortException = _abortException;
+                    _abortException = null;
+
+                    // There is an inherent race between receive and close. Removing the last message from the channel
+                    // makes Input.Completion task completed and runs this continuation. We need to await _receiveLoopTask
+                    // to make sure that the message removed from the channel is processed before we drain the queue.
+                    // There is a short window between we start the channel and assign the _receiveLoopTask a value.
+                    // To make sure that _receiveLoopTask can be awaited (i.e. is not null) we need to await _startTask.
+                    _logger.ProcessRemainingMessages(_connectionId);
+
+                    await _startTcs.Task;
+                    await _receiveLoopTask;
+
+                    _logger.DrainEvents(_connectionId);
+
+                    await Task.WhenAny(_eventQueue.Drain().NoThrow(), Task.Delay(_eventQueueDrainTimeout));
+
+                    _logger.CompleteClosed(_connectionId);
+
+                    // At this point the connection can be either in the Connected or Disposed state. The state should be changed
+                    // to the Disconnected state only if it was in the Connected state.
+                    ChangeState(from: ConnectionState.Connected, to: ConnectionState.Disconnected);
+
+                    try
+                    {
+                        if (t.IsFaulted)
+                        {
+                            _closedEvent.Invoke(t.Exception.InnerException);
+                        }
+                        else
+                        {
+                            // Call the closed event. If there was an abort exception, it will be flowed forward
+                            // However, if there wasn't, this will just be null and we're good
+                            _closedEvent.Invoke(abortException);
+                        }
+                    }
+                    catch (Exception ex)
+                    {
+                        // Suppress (but log) the exception, this is user code
+                        _logger.ErrorDuringClosedEvent(ex);
+                    }
+
+                    _closeTcs.SetResult(null);
```

Comment:
FYI, I moved this to be after `_closedEvent` is called, this means closed callbacks will finish before `StopAsync` returns